### PR TITLE
code review

### PR DESCRIPTION
**Pull request by RonKant**

### CODE REVIEW

File: data/handle_data.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    padded_df = torch.tensor(len(xtrain_seq_premise),2,len(xtrain_seq_premise[0]))
+    padded_df[0] = xtrain_seq_premise
+    padded_df[1] = xtrain_seq_hypothesis
+    features_train, features_test, targets_train, targets_test = train_test_split(padded_df,
+                                                                                target,
+                                                                                test_size = 0.95,
+                                                                                random_state = 42) 
+
+    # Wait, is this a CPU tensor now? Why? Where is .to(device)?
```
Comment (RonKant) [inline data/handle_data.py:119]: everything below this line is code duplicated twice for train / test. If it's the exact same process, then it should be defined once, in another function, and only applied twice as a function call.

File: data/handle_data.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        return (self.x[index], self.y[index])
+
+    def __len__(self):
+        return len(self.x)
+    
+
+def create_datasets(xtrain_seq_premise, xtrain_seq_hypothesis,target):
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    padded_df = torch.tensor(len(xtrain_seq_premise),2,len(xtrain_seq_premise[0]))
```
Comment (RonKant) [inline data/handle_data.py]: What is this object? what does it represent/contain?
lines 97-100 can be simplified

File: experiments/simple_train_loop.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import numpy as np
+
+def simple_train_loop(train_loader, optimizer,criterion,device,model,split_val,num_of_epochs):
+
+    losses = []
+    losses_plt = []
+
+    train_step_fuc = train_step.simple_train_step(optimizer,criterion,device,model,split_val)
+
+    for epoch in range(num_of_epochs):
```
Comment (RonKant) [inline experiments/simple_train_loop.py:11]: This part below is great, and indeed pretty simple.
the train_step_fuc part (Above), however, is too complex:

why do you need to create and return a function dynamically, in every run, according to optimizer, criterion, device, etc.

a more simple, readable solution would be to have train_step be a fixed function, that also receives additional parameter.

think if you can find an even cleaner solution, that does not require you to pass a bunch of parameters around between functions, and still have a single fixed training step function. (Hint: classes).

File: experiments/simple_train_step.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from itertools import chain
+import torch
+
+
+def simple_train_step(optimizer,criterion,device,model,split_val):
+        
+    def train_step(x, y):
+        model.to(device)
```
Comment (RonKant) [inline experiments/simple_train_step.py:8]: this does not need to happen every step -> extract this to a higher layer in the logic

File: experiments/simple_train_step.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        premise, hypothesis = torch.split(x,split_val,dim=1)
+        
+        first_hidden_pre = torch.rand_like(premise).to(device)
```
Comment (RonKant) [inline experiments/simple_train_step.py:11]: these 4 lines seem like a default logic that is more related to the model / data than to your training step logic.
You want to keep your training logic as pure as possible, to improve readability and prevent change amplification (in case this default data processing has to change).
Can you think of a more fitting place for this?

File: inference_levels/InputEncoders/ESIM_encoder.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import torch
+from .input_encoding import BaseInputEncoder
+
+class ESIMInputEncoder(BaseInputEncoder):
```
Comment (RonKant) [inline inference_levels/InputEncoders/ESIM_encoder.py:4]: Besides the issue of being a duplication of BaseInputEncoder (without adding too much significant behavior), this class is not necessary:
logically, all it does is have some encoding_layer, which is applied twice, and that's it.
Instead of creating a class for this (which is more OOP but increases cognitive load, consider extracting this behaviour to your model class

`
def __init__(...):
self.encoding_layer = ... # LSTM is passed from outside

def __forward__(...):
  a_hat = self.encoding_layer(premise, h0, c0)
 b_hat = self.encoding_layer(hypothesis, h0, c0)

`

It is true that generally, we'd want to have a nice class/module for each separate entity, such as an "encoding step", however -- when the resulting classes are very shallow (have a small implementation and a large interface), then it is not worth it. You can read more about it In Chapter 4 of 'a Philosophy of Software Design'

File: inference_levels/local_inferencers/local_inference.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import torch
+
+
+class BaseLocalInference(torch.nn.Module):
+    def __init__(self,embedding_size):
```
Comment (RonKant) [inline inference_levels/local_inferencers/local_inference.py:5]: embedding_size is unused

File: inference_levels/local_inferencers/local_inference.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        super().__init__()
+
+        self.energy_metrix = None
```
Comment (RonKant) [inline inference_levels/local_inferencers/local_inference.py:8]: does this have to be a field? If it is re-created in each forward call, then it is better to have it a local variable in that function.
Saving it to the object state might be confusing, since it implies that this field has a purpose in the future, for which it is stored, and it might cause bugs when you try to access this field at inappropriate times.

File: inference_levels/local_inferencers/local_inference.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+
+        self.zero_dim_softmax = torch.nn.Softmax(0)
```
Comment (RonKant) [inline inference_levels/local_inferencers/local_inference.py]: these fields can also be removed.
you can use torch.nn.functional.softmax instead, as a function -- there is not reason to have this as a separate object (it only increases cognitive, since reading 'torch.nn.functional.softmax(stuff) does not require looking at class initialization to understand where self.zero_dim_softmax comes from)

File: inference_levels/predictors/predictor.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import torch
+
+
+class BasePredictor(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+
+    def get_avg_max_vector(self,vector):
+        avg_vec = torch.sum(vector)/len(vector)
```
Comment (RonKant) [inline inference_levels/predictors/predictor.py]: please simplify

File: train_ESIM.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+train_loader = DataLoader(dataset=train_data, batch_size=1, shuffle=True)
+
+input_size,hidden_size = 3,3
+input_encoding = ESIMInputEncoder(input_size,hidden_size)
+local_inference = BaseLocalInference(hidden_size)
+composition_layer = torch.nn.LSTM(input_size, hidden_size, num_layers=1, batch_first=False, dropout=0.0, bidirectional=True)
+infrence_composition = BaseInferenceComposition(hidden_size,composition_layer)
+predictor = BasePredictor()
+model = ESIMModel(input_encoding,local_inference,infrence_composition,predictor)
+train_exp.run_simple_train_exp(train_loader,model)
```
Comment (RonKant) [inline train_ESIM.py]: lets enhance your training "main" and make it easier to use for running experiments.
Lets say, I want to run the same code, but with different batch size, to test some hypothesis I have about my model.

It would be lovely if I could run it from my terminal without changing the code, i.e.:

python train_ESIM.py --batch_size 32
python train_ESIM.py --batch_size 64
python train_ESIM.py --batch_size 128

read about argparse, and see an example in https://github.com/AI-COE-git/Accent_Conversion/blob/master/train_diffwave.py

Think about which variable make sense as arguments, and try to implement your main script to allow this type of flexibility
